site/app/libraries/database/DatabaseQueries.php

```diff
@@ -1126,6 +1126,52 @@ public function getGroupForUserInClass($semester, $course_name, $user_id) {
         return intval($this->submitty_db->row()['user_group']);
     }
 
+    /**
+     * Gets an unique array of active users in courses_users, active users are users enrolled in a course in the current semester
+     * Allows the filtering the group of users returned by this function
+     *
+     * @param bool $instructor To include instructors or not
+     * @param bool $fullAccess To include full access graders or not
+     * @param bool $limitedAccess To include limited access graders or not
+     * @param bool $student To include students or not
+     * @param bool $faculty to include faculty level users or not
+     * @return array - array of userids active in the specified semester
+     */
+    public function getActiveUserIds(bool $instructor, bool $fullAccess, bool $limitedAccess, bool $student, bool $faculty): array {
+        $result_rows = [];
+        $this->submitty_db->query(
+            "SELECT courses_users.user_id as user_id, courses_users.user_group as user_group
+                FROM courses_users LEFT JOIN courses 
+                ON courses_users.semester = courses.semester AND courses_users.course = courses.course
+                WHERE courses.status = 1"
+        );
+        $results = $this->submitty_db->rows();
+        foreach ($results as $row) {
+            if ($instructor && $row["user_group"] == 1) {
+                $result_rows[] = $row["user_id"];
+            }
+            elseif ($fullAccess && $row["user_group"] == 2) {
+                $result_rows[] = $row["user_id"];
+            }
+            elseif ($limitedAccess && $row["user_group"] == 3) {
+                $result_rows[] = $row["user_id"];
+            }
+            elseif ($student && $row["user_group"] == 4) {
+                $result_rows[] = $row["user_id"];
+            }
+        }
+        if ($faculty) {
+            $this->submitty_db->query(
+                "SELECT user_id FROM users WHERE user_access_level <= 2"
+            );
+            $results = $this->submitty_db->rows();
+            foreach ($results as $row) {
+                $result_rows[] = $row["user_id"];
+            }
+        }
```

Comment:
What's the efficiency of this query for a very large number of users, especially when run multiple times and iterated through when showing the email page?

Is it cheaper to do all of the logic user group comparison logic + duplicate removal within the DB, like:

```php
        $args = [];
        $extra_join = '';
        $extra_where = '';
        if ($instructor) {
            $args[] = '1';
        }
        if ($fullAccess) {
            $args[] = '2';
        }
        if ($limitedAccess) {
            $args[] = '3';
        }
        if ($student) {
            $args[] = '4';
        }
        if ($faculty) {
            $extra_join = ' INNER JOIN users ON courses_users.user_id = users.user_id ';
            $extra_where = ' OR users.user_access_level <= 2';
        }
        $result_rows = [];
        $this->submitty_db->query(
            "SELECT DISTINCT courses_users.user_id as user_id
                FROM courses_users INNER JOIN courses 
                ON courses_users.semester = courses.semester AND courses_users.course = courses.course
                " . $extra_join . "
                WHERE courses.status = 1 AND user_group IN (" . implode(', ', $args) . ")" . $extra_where
        );
        return array_map(function ($row) { return $row['user_id']; }, $this->submitty_db->rows());
```

Though, also, could remove the `array_map in favor of modifying where the results are used to just use the associative array.

Reply:
Although I have very little knowledge of databases, I would imagine that my solution will be very inefficient for large amount of users, and will likely be much cheaper to use your solution. I have, however, modified `args` to initialize to `[-1]`, which is an invalid group in order to avoid a situation where the query becomes invalid when only faculty is selected. Is this a bad way to work around this situation?